### ISSUE TITLE
PHP Fatal error:  Uncaught GuzzleHttp\Exception\RequestException: cURL error 2: easy handle already used in multi handle

### ISSUE THREAD

**Issue by EsSa55** [guzzle/guzzle#2617]
Guzzle:  6.2.3
PHP:      7.3.16
Curl:      7.19.7

**Description**  

Library used for Google Shopping API product feeds

Feeds worked successfully every 20 minutes for 2 years and first 5 hours of 17 April 2020

Then:

PHP Fatal error:  Uncaught GuzzleHttp\Exception\RequestException: cURL error 2: easy handle already used in multi handle (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) in ../vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:162
Stack trace:
#0 ../vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(129): GuzzleHttp\Handler\CurlFactory::createRejection(Object(GuzzleHttp\Handler\EasyHandle), Array)
#1 ../vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(89): GuzzleHttp\Handler\CurlFactory::finishError(Object(GuzzleHttp\Handler\CurlHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory))
#2 ../vendor/guzzlehttp/guzzle/src/Handler/CurlHandler.php(43): GuzzleHttp\Handler\CurlFactory::finish(Object(GuzzleHttp\Handler\CurlHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory))
#3 ../vendor/guzzlehttp/guzzle/src/Handl in ../vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 162

No change to server, site configuration or any associated php or library files.

Same error every time the script now runs.

I am not a programmer but can follow instructions!

Any leads or guidance will be appreciated as the feed developer is no longer available.

TIA

**Comment by EsSa55** [comment]
Updated curl to v7.69.1. Issue persists

**Comment by GrahamCampbell** [comment]
> Updated curl to v7.69.1. Issue persists

What do you mean by this? The curl cli binary you might have on your machine is not related to lib-curl.

**Comment by EsSa55** [comment]
Sorry, by 'issue persists' I mean the error is the same.

There is a possible solution on [stackoverflow](https://stackoverflow.com/questions/61285075/php-fatal-error-uncaught-guzzlehttp-exception-requestexception-curl-error-2-e?noredirect=1#comment108417115_61285075) but it applies to Stripe and I have no idea how to implement it in Guzzle

**Comment by GrahamCampbell** [comment]
I mean what did you mean by updated curl?

**Comment by EsSa55** [comment]
Upgraded the curl version

**Comment by GrahamCampbell** [comment]
What do you see when you run `php -r 'echo curl_version()["version"];'`.

**Comment by EsSa55** [comment]
7.69.1

**Comment by ingberrio** [comment]
I get the same problem, GuzzleHttp\Exception\RequestException: cURL error 2:, this appear when on godaddy service active ssl with some browser not working (safari, firefox), I`ve the same configuration

**Comment by pavelgajdos** [comment]
It helped me to configure the client as follows:

```
$client = new Client([
    'handler' => new CurlHandler([
         'handle_factory' => new CurlFactory(0)
    ])
]);
```

However, I still think there is a bug somewhere. It stopped working on its own, nothing has changed neither in the app or on the server.

**Comment by capuderg** [comment]
Hi,

some of our users are also reporting this issue: https://wordpress.org/support/topic/curl-error-2-easy-handle-already-used-in-multi-handle/

In our WP plugin: WP Mail SMTP we are using "google/apiclient" lib and it requires guzzle.

I can't reproduce this issue on my shared hosting or locally.

There were no changes to the plugin code when this issue started appearing for our users.

@EsSa55 When you say "All is well again", does that mean that the error is no longer present for you? Did cURL update on your server when these errors started to appear? Did you downgrade to an older version of cURL in order to resolve this issue?

**Comment by EsSa55** [comment]
@capuderg That's right. The error is no longer present, having requested my server host to downgrade cURL + libcurl to the previous version.
cURL did update automatically on my server. Sorry, I don't actually know what the new version was.

# curl -V
curl 7.69.1 (x86_64-redhat-linux-gnu) libcurl/7.69.1 OpenSSL/1.0.1e-fips zlib/1.2.3 c-ares/1.16.0 libssh2/1.8.2
Release-Date: 2020-03-11
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS GSS-API HTTPS-proxy IPv6 Kerberos Largefile libz Metalink NTLM NTLM_WB SPNEGO SSL UnixSockets

**Comment by ingberrio** [comment]
> It helped me to configure the client as follows:
> 
> ```
> $client = new Client([
>     'handler' => new CurlHandler([
>          'handle_factory' => new CurlFactory(0)
>     ])
> ]);
> ```
> 
> However, I still think there is a bug somewhere. It stopped working on its own, nothing has changed neither in the app or on the server.
> 
> ### Edit
> I was told that there may had been an automatic update as we use cPanel on that server. Unfortunately, I don't know more.

Before a long days my fix, replace StreamHandler instead of CurlHandler

**Comment by EsSa55** [comment]
Posted by @Matt on https://stackoverflow.com/questions/61285075/php-fatal-error-uncaught-guzzlehttp-exception-requestexception-curl-error-2-e?noredirect=1#comment108426410_61285075

Temporary fix:

Just in case anyone reading this is using Laravel. We suddenly started having the same problem a few days ago, tried installing different cURL versions and setting cURL options, nothing worked. I fixed it by changing the vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php file. Look for the line that says

if (count($this->handles) >= $this->maxHandles) {
curl_close($resource);
} else {
...
}
Comment this all out, and instead of the if/else just write

curl_close($resource);

In other words no matter what the handles count is you always close the cURL connection. This worked instantly for us!

I thought the cURL version downgrade was the 'fix', but it worked for just 2 days.

**Comment by pavelgajdos** [comment]
@EsSa55 I got inspired by this solution but instead of commenting out packaged lines of code I kind of forced the condition to be always true, therefore closing a resource every time. In my previous comment the number 0 is assigned to `$this->maxHandles`

**Comment by EsSa55** [comment]
@pavelgajdos As a non-programmer, I was able to follow the instruction re specific commenting out. However, I am lost as to where/how your solution is applied re 'configure the client'.

**Comment by pavelgajdos** [comment]
@EsSa55 In this case you are better off with commenting the lines out even though it is not as clean solution as mine or @ingberrio's :-) 

Applying my solution heavily depends on your specific software and a guide, which would cover most possible situations in your code, would be too long and unnecessary for this issue.

Anyway, I hope we will know a cause of this bug or there will be a hotfix soon.

**Comment by GrahamCampbell** [comment]
> curl -V

This does not tell you the version of libcurl PHP is using.

**Comment by sandervanhooft** [comment]
Can someone confirm this is an issue caused in this package?

I’m getting reports of this issue in one of my packages that requires guzzle, but cannot reproduce it.

**Comment by Gaffen** [comment]
This issue is also being encountered in Stripes' php client, with reference to guzzle:
https://github.com/stripe/stripe-php/issues/918

**Comment by Gaffen** [comment]
I also believe this issue is being encountered with https://github.com/awesomemotive/WP-Mail-SMTP, for their gmail integration which also uses guzzle.

**Comment by gmponos** [comment]
From https://curl.haxx.se/libcurl/c/libcurl-errors.html
>CURLE_FAILED_INIT (2)
> Very early initialization code failed. This is likely to be an internal error or problem, or a resource problem where something fundamental couldn't get done at init time. 

and from https://cdcvs.fnal.gov/redmine/projects/simons-dm/wiki/Curl_Error_Codes
> If we ever get this from our probe script, it means we have a broken version of curl on the system we are using, which can't find its own parts. So the user installed their own curl and it didn't build correctly, etc. Have them check their PATH command search path etc. installed packages, etc.


Also from the original author:

> Guzzle: 6.2.3
> PHP Fatal error: Uncaught GuzzleHttp\Exception\RequestException: cURL error 2: easy handle already used in multi handle (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) in ../vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:162

So now let's head there: https://github.com/guzzle/guzzle/blob/6.2.3/src/Handler/CurlFactory.php#L162 

🤔

**Comment by gmponos** [comment]
The only versions that I found this line match is 6.0.0 and 6.0.1 https://github.com/guzzle/guzzle/blob/6.0.1/src/Handler/CurlFactory.php#L162

Please check your setup and use and updated version.

**Comment by Gaffen** [comment]
@gmponos thanks for this. Guzzle is embedded as a dependency so I don't fully have access over this, however it does at least help isolate the problem.

I'll investigate the versions bundles with the two plugins that have it as a dependency and confirm. If it turns out the guzzle version is the issue and not guzzle itself it could mean this issue can be closed.

**Comment by capuderg** [comment]
In the WP Mail SMTP plugin the lowest Guzzle version that was used for a couple of releases now was '6.4.1'. Even when this issue started to occur for our users the plugin used that version...

Here are some updates from the users experiencing this issue: https://wordpress.org/support/topic/curl-error-2-easy-handle-already-used-in-multi-handle/page/3/

A lot of users reported they are hosting on GoDaddy. Is this maybe a GoDaddy specific issue?

**Comment by Gaffen** [comment]
I'm pretty sure it's not godaddy; I think it might be something to do with CentOS (I've seen other centOS users report similar things - do you know if your godaddy users are on CentOS systems @capuderg ?)

**Comment by gmponos** [comment]
Firstly I can't safely exclude that this couldn't be a guzzles issue.. but I would like to have a stack trace that I can trust.

Secondly (not a WP expert) but unless you are using wppackagist plugins include all their vendor in the install right? How does the autoloading behaves when two plugins include in their vendors `Guzzle`? Maybe another plugin is using outdated guzzle version and it's the reason we have this stack trace? 

If this "could" be an issue then what is the composer version used to generate the autoloading files for each version of you WP plugin?

**Comment by capuderg** [comment]
@Gaffen I don't have any additional info on their systems, apart from the info on the forum post I pasted above. Sorry.

**Comment by capuderg** [comment]
@gmponos If two WP plugins use the same composer package then the plugin that calls the autoload first will "register" the package and the second plugin's package will be ignored...

I've asked a user to list all plugins that he uses and verified if there is some kind of conflict like this, but there was none.

**Comment by amitshahc** [comment]
I am facing the same issue. I can't downgrade the things on the server. when we deployed the project the socialite plugin was working perfectly. suddenly it started throwing above error in the logs. please provide the solution asap. i also don't prefer to change in vendor files.

**Comment by Sensetivity** [comment]
I also facing this problem. 

PHP 5.6.40 (cli) (built: Jun 15 2020 20:46:06)
cURL 7.69.1

> Exception 'Google\Cloud\Core\Exception\ServiceException' with message 'cURL error 2: easy handle already used in multi handle (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)'

**Comment by amitshahc** [comment]
> Posted by @matt on https://stackoverflow.com/questions/61285075/php-fatal-error-uncaught-guzzlehttp-exception-requestexception-curl-error-2-e?noredirect=1#comment108426410_61285075
> 
> Temporary fix:
> 
> Just in case anyone reading this is using Laravel. We suddenly started having the same problem a few days ago, tried installing different cURL versions and setting cURL options, nothing worked. I fixed it by changing the vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php file. Look for the line that says
> 
> if (count($this->handles) >= $this->maxHandles) {
> curl_close($resource);
> } else {
> ...
> }
> Comment this all out, and instead of the if/else just write
> 
> curl_close($resource);
> 
> In other words no matter what the handles count is you always close the cURL connection. This worked instantly for us!
> 
> I thought the cURL version downgrade was the 'fix', but it worked for just 2 days.

I had to use the temporary fix mentioned here. but I don't recommend this and the developer should come up with the solution asap.

**Comment by jamesozzie** [comment]
Downgrading the libcurl version resolved a similar issue for a Site Kit plugin user, as reported [here](https://wordpress.org/support/topic/error-curl-error-2/#post-13237544).

**Comment by irondan** [comment]
I'm in the (unique?) position where I'm managing a handful of Drupal sites. This error prevents some of them from retrieving information about project updates (the core Update module in Drupal 8/9 uses Guzzle to retrieve this information). Here's what I'm seeing:

Test | Drupal Version | PHP Version | cURL Version | Fails With error 2
--- | --- | --- | --- | ---
Site 1 Prod | 8.9.3 | 7.3.21 | 7.71.1 | No
Site 1 Stage | 8.9.3 | 7.3.20 | 7.71.0 | No
Site 2 Prod | 8.9.2 | 7.2.33 | 7.71.1 | No
Site 2 Stage | 8.9.2 | 7.2.33 | **7.69.1** | **Yes**
Site 3 Prod | 8.9.3 | 7.3.14 | 7.29.0 | No
Site 3 Stage | 8.9.3 | 7.3.21 | **7.69.1** | **Yes**
Site 4 | 9.0.3 | 7.4.8 | 7.29.0 | No

Far from conclusive information to be sure, but it's looking like this issue might be version specific to cURL. I will try updating that specifically any come back with any results.

**Comment by Nyholm** [comment]
As @gmponos shows, it is likely a curl issue. 

I’m closing this since nobody has shown a stack trace, a reproducible script or a reliable version of guzzle where this happens. Feel free to open a new issue if you can provide a short reproducible script.

**Comment by ItsMeStevieG** [comment]
I am still getting this while using 2 packages (Packagist libraries) from different developers that have guzzle as a dependency.

I am getting this using the Following:

**CURL:**
`curl 7.74.0 (x86_64-apple-darwin20.2.0) libcurl/7.74.0 (SecureTransport) OpenSSL/1.1.1i zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.0 libssh2/1.9.0 nghttp2/1.42.0 librtmp/2.3
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz Metalink MultiSSL NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets zstd`

**Guzzle**
`guzzlehttp/guzzle (7.2.0)`

**Stack Trace:**
`Fatal error: Uncaught GuzzleHttp\Exception\RequestException: cURL error 2: (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://oauth.nitrado.net/oauth/v2/token in /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:211 Stack trace: #0 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(158): GuzzleHttp\Handler\CurlFactory::createRejection(Object(GuzzleHttp\Handler\EasyHandle), Array) #1 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(110): GuzzleHttp\Handler\CurlFactory::finishError(Object(GuzzleHttp\Handler\CurlHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory)) #2 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/CurlHandler.php(47): GuzzleHttp\Handler\CurlFactory::finish(Object(GuzzleHttp\Handler\CurlHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory)) #3 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(28): GuzzleHttp\Handler\CurlHandler->__invoke(Object(GuzzleHttp\Psr7\Request), Array) #4 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(48): GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}(Object(GuzzleHttp\Psr7\Request), Array) #5 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php(64): GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}(Object(GuzzleHttp\Psr7\Request), Array) #6 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Middleware.php(31): GuzzleHttp\PrepareBodyMiddleware->__invoke(Object(GuzzleHttp\Psr7\Request), Array) #7 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/RedirectMiddleware.php(71): GuzzleHttp\Middleware::GuzzleHttp\{closure}(Object(GuzzleHttp\Psr7\Request), Array) #8 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Middleware.php(63): GuzzleHttp\RedirectMiddleware->__invoke(Object(GuzzleHttp\Psr7\Request), Array) #9 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/HandlerStack.php(75): GuzzleHttp\Middleware::GuzzleHttp\{closure}(Object(GuzzleHttp\Psr7\Request), Array) #10 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Client.php(331): GuzzleHttp\HandlerStack->__invoke(Object(GuzzleHttp\Psr7\Request), Array) #11 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Client.php(107): GuzzleHttp\Client->transfer(Object(GuzzleHttp\Psr7\Request), Array) #12 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Client.php(123): GuzzleHttp\Client->sendAsync(Object(GuzzleHttp\Psr7\Request), Array) #13 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/league/oauth2-client/src/Provider/AbstractProvider.php(608): GuzzleHttp\Client->send(Object(GuzzleHttp\Psr7\Request)) #14 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/league/oauth2-client/src/Provider/AbstractProvider.php(621): League\OAuth2\Client\Provider\AbstractProvider->getResponse(Object(GuzzleHttp\Psr7\Request)) #15 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/league/oauth2-client/src/Provider/AbstractProvider.php(537): League\OAuth2\Client\Provider\AbstractProvider->getParsedResponse(Object(GuzzleHttp\Psr7\Request)) #16 /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/test.php(45): League\OAuth2\Client\Provider\AbstractProvider->getAccessToken(Object(League\OAuth2\Client\Grant\AuthorizationCode), Array) #17 {main} thrown in /Users/itsmestevieg/Sites/stevieg.online/oauth2-nitrado/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 211`

Any help would be greatly appreciated.

**Comment by xgaibtcgd** [comment]
Why would this not run for my admin user that falls in the allowed access group 6055?

<pre>
<?php
/*Load up wordpress externally*/
define( 'WP_USE_THEMES', false );
define( 'COOKIE_DOMAIN', false );
define( 'DISABLE_WP_CRON', true );
require('/opt/bitnami/wordpress/wp-load.php');

//DB
$servername = "localhost";
$username = "node";
$password = "n0d3p4ss";
$dbname = "dashboard";

//ops
function ops($op, $a, $b){
  switch($op){
    case '=':
      return $a == $b;
    break;
    case '>':
      return $a > $b;
    break;
    case '<':
      return $a < $b;
    break;
    case '>=':
      return $a >= $b;
    break;
    case '<=':
      return $a <= $b;
    break;
  }
}

/**
 * Check if user has access
 *
 * @param $uid
 *
 * @return bool
 */
if( ! function_exists( 'x_authorize' ) ) {
  function x_authorize($uid){

    $found = false;
    $pass = [6079,6055,5923];
    //Confirm membership active
    $memberships = wc_memberships_get_user_active_memberships( $uid );

    foreach($memberships as $ship){
      if(in_array($ship->plan_id, $pass)) $found = true;
    }

    return $found;

  }
}

// Create connection
$conn = new mysqli($servername, $username, $password, $dbname);

// Check connection
if ($conn->connect_error) {
  die("Connection failed: " . $conn->connect_error);
}
line();
echo date("Y-m-d h:i:sa").PHP_EOL;
echo "XABCD Chat Notification Script".PHP_EOL;
echo "INIT :: DB Connected successfully".PHP_EOL;

//Users
$users = get_users();


//Chat
$result = $conn->query("SELECT * FROM chat order by id DESC LIMIT 100");
$allchat = $result->fetch_all(MYSQLI_ASSOC);
$chat = array_filter($allchat, function($message){  if($message['private'] == 'all') return true; });

echo 'Loaded '.count($chat).' Chat Messages'.PHP_EOL;
$conn->close();

//RUN!!
foreach($users as $user){
  if( x_authorize( $user->id ) ) {
    $emailAlerts = json_decode(get_user_meta($user->id,'emailAlerts', true));
    runChatNotifications($chat, $user, $emailAlerts, $allchat);
  }
}


//Helper functions
function runChatNotifications($chat, $user, $emailAlerts, $allchat){

  line();
  print_r('Running chat for '.$user->ID.' '.$user->data->user_nicename.' '.$user->data->user_email.PHP_EOL);

  if($emailAlerts->chat == 1 && isUserAlertAllowed( $user, $emailAlerts ) ){

    $lastChatId = get_user_meta($user->ID, 'lastChatId', true);
    $currentChat = array_shift($chat);
    print_r('Current Chat is: #'.$currentChat['id'].PHP_EOL.'Last Chat Seen: #'.$lastChatId.PHP_EOL);
    if($currentChat['id'] > $lastChatId) {
      sendChatAlert($user, $currentChat);
      increaseDailyAlertCounter( $user );
    }
  }

  if($emailAlerts->private == 1 && isUserAlertAllowed( $user, $emailAlerts ) ){

    $lastPrivateId = get_user_meta($user->ID, 'lastPrivateId', true);
    $private = array_filter($allchat, function($message) use ($user){  if($message['private'] == $user->data->user_nicename) return true; });
    $currentPrivate = array_shift($private);

    print_r('Current Private is: #'.$currentPrivate['id'].PHP_EOL.'Last Private Seen: #'.$lastPrivateId.PHP_EOL);
    if($currentPrivate['id'] > $lastPrivateId) {
      sendPrivateAlert($user, $currentPrivate);
      increaseDailyAlertCounter( $user );
    }
  }

  if( $emailAlerts->mention == 1 && isUserAlertAllowed( $user, $emailAlerts ) ){

    $lastMentionId  = get_user_meta($user->ID, 'lastMentionId', true);
    if( ! $lastMentionId || $lastMentionId === '0' ) {
      return;
    }

    $mention        = array_filter($allchat, function($message) use ($lastMentionId){  if( $message['id'] == $lastMentionId ) return true; });
    $currentMention = array_shift($mention);

    print_r('Current Mention is: #'.$currentMention['id'].PHP_EOL);
    sendMentionAlert($user, $currentMention);
    increaseDailyAlertCounter( $user );
  }
}


function sendChatAlert($user, $currentChat){
  $to = $user->data->user_email;
  $subject = 'Chat Alert Notification';
  $body = '
<p>Hello,
<p>There are people chatting inside XABCD Trading, you might want to join in for the discussion.<br>
<p><a href="https://www.xabcdtrading.com/dashboard/">https://www.xabcdtrading.com/dashboard/</a></p>
<p>- XABCD Team<br>
Turn <a href="https://www.xabcdtrading.com/members-education-center/faq/">on/off</a> email alerts.</p>
  ';
  $headers = array('Content-Type: text/html; charset=UTF-8');
  if(wp_mail( $to, $subject, $body, $headers )){
    update_user_meta($user->ID, 'lastChatId', $currentChat['id']);
    print_r('Sent Chat Alert for #'.$currentChat['id'].PHP_EOL);
  }
}

function sendPrivateAlert($user, $currentPrivate){
  $to = $user->data->user_email;
  $subject = 'Private Message Alert Notification';
  $body = '
  <p>Hello,</p>
  <p>You have a new private message that was sent to you. Login to see who is talking to you.</p>
  <p><a href="https://www.xabcdtrading.com/dashboard/">https://www.xabcdtrading.com/dashboard/</a></p>
  <p>- XABCD Team<br>
  Turn <a href="https://www.xabcdtrading.com/members-education-center/faq/">on/off</a> email alerts.</p>
  ';
  $headers = array('Content-Type: text/html; charset=UTF-8');
  if(wp_mail( $to, $subject, $body, $headers )){
    update_user_meta($user->ID, 'lastPrivateId', $currentPrivate['id']);
    print_r('Sent Private Chat Alert!'.PHP_EOL);
  }
}

function sendMentionAlert($user, $currentMention){
  $to = $user->data->user_email;
  $subject = 'Mention Message Notification';
  $body = '
  <p>Hello,</p>
  <p>You have been mentioned in the message.</p>
  <p><a href="https://www.xabcdtrading.com/dashboard/">https://www.xabcdtrading.com/dashboard/</a></p>
  <p>- XABCD Team<br>
  Note:  Go here to learn how to manage notifications: <a href="https://www.xabcdtrading.com/how-to-notifications/">https://www.xabcdtrading.com/how-to-notifications/</a></p>
  ';
  $headers = array('Content-Type: text/html; charset=UTF-8');
  if(wp_mail( $to, $subject, $body, $headers )){
    update_user_meta($user->ID, 'lastMentionId', 0 );
    print_r('Sent Mention Alert!'.PHP_EOL);
  }
}

function line(){
  print_r('******************************************************************'.PHP_EOL);
}

/**
 * Validation if user has daily limit to send him Alert.
 *
 * @param $user
 * @param $emailAlerts
 *
 * @return bool
 */
function isUserAlertAllowed( $user, $emailAlerts ) {

  if( ! isset( $emailAlerts->perDay ) ) {
    return true;
  }

  return $emailAlerts->perDay > get_user_meta( $user->id, 'emails_day_counter', true );
}

/**
 * Increase daily email Counter
 *
 * @param $user
 *
 * @return int
 */
function increaseDailyAlertCounter( $user ) {
  $counter = get_user_meta( $user->id, 'emails_day_counter', true );
  update_user_meta( $user->id, 'emails_day_counter', intval( $counter ) + 1 );

  return intval( $counter ) + 1;
}
